### PR TITLE
[GSB] Add verification of all generic signatures within a module. 

### PR DESCRIPTION
**Pull request by DougGregor**
Add a verification pass to ensure that all of the generic signatures in
a module are both minimal and canonical. The approach taken is quite
direct: for every (canonical) generic signature in the module, try
removing a single requirement and forming a new generic signature from
the result. If that new generic signature that provide the removed
requirement, then the original signature was not minimal.

Also canonicalize each resulting signature, to ensure that it meets the
requirements for a canonical signature.

Add a test to ensure that all of the generic signatures in the Swift
module are minimal and canonical, since they are ABI.

Resolves [SR-3733](https://bugs.swift.org/browse/SR-3733) / rdar://problem/31412994

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: test/Generics/validate_stdlib_generic_signatures.swift [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// Verifies that all of the generic signatures in the standard library are
+// minimal and canonical.
+
+// RUN: %target-typecheck-verify-swift -typecheck %s -verify-generic-signatures Swift
```
Comment (slavapestov) [inline test/Generics/validate_stdlib_generic_signatures.swift]: So this passes now? :)

Comment (DougGregor) [reply test/Generics/validate_stdlib_generic_signatures.swift]: YES!

Comment (DougGregor) [reply test/Generics/validate_stdlib_generic_signatures.swift]: My comfort level in GSB minimization for ABI stability has increased 100x

**Pull request merged**